File: mcp-client-typescript/oci_mcp_client.ts
import { GenerativeAiInferenceClient, models, requests } from "oci-generativeaiinference";
import {
  InstancePrincipalsAuthenticationDetailsProviderBuilder,
  NoRetryConfigurationDetails
} from "oci-common";
import { env } from "process";

import { Client } from "@modelcontextprotocol/sdk/client/index.js";
import { StdioClientTransport } from "@modelcontextprotocol/sdk/client/stdio.js";
import readline from "readline/promises";

const servingMode: models.OnDemandServingMode = {
    modelId: env.TF_VAR_genai_cohere_model,
    servingType: "ON_DEMAND",
};

class MCPClient {
  private mcp: Client;
  private llm: GenerativeAiInferenceClient = null;
  private transport: StdioClientTransport | null = null;
  private tools: Object[] = [];

  constructor() {
    this.mcp = new Client({ name: "mcp-client-cli", version: "1.0.0" });
  }

  async initLLM() {
    const provider = await new InstancePrincipalsAuthenticationDetailsProviderBuilder().build();
    this.llm = new GenerativeAiInferenceClient({
        authenticationDetailsProvider: provider,
      }
    );
    this.llm.endpoint = "https://inference.generativeai."+env.TF_VAR_region+".oci.oraclecloud.com";
  }

  async connectToServer(serverScriptPath: string) {
    /**
     * Connect to an MCP server
     *
     * @param serverScriptPath - Path to the server script (.py or .js)
     */
    try {
      // Determine script type and appropriate command
      const isJs = serverScriptPath.endsWith(".js");
      const isPy = serverScriptPath.endsWith(".py");
      if (!isJs && !isPy) {
        throw new Error("Server script must be a .js or .py file");
      }
      const command = isPy
        ? process.platform === "win32"
          ? "python"
          : "python3.12"
        : process.execPath;

      // Initialize transport and connect to server
      this.transport = new StdioClientTransport({
        command,
        args: [serverScriptPath],
      });
      this.mcp.connect(this.transport);
      console.log("before ListTools"); 
      await new Promise(r => setTimeout(r, 2000));

      // List available tools
      const toolsResult = await this.mcp.listTools();

      console.log( "toolsResult", JSON.stringify(toolsResult) );      
      this.tools = toolsResult.tools.map((tool) => {
        console.log( "tool.inputSchema", JSON.stringify(tool.inputSchema) );  
        var tool_schema = tool.inputSchema.properties;
        console.log( "tool_schema", JSON.stringify(tool_schema) );  
        var params = {}
        Object.keys(tool_schema).forEach(function(key,index) {
            params[key] = {
              type: tool_schema[key].type,
              description: tool_schema[key].name,
              isRequired: false
            }
        });
        console.log( "tool.inputSchema", JSON.stringify(params) );      
        return {
          name: tool.name,
          description: tool.description,
          parameterDefinitions: params,
        };
      });
      console.log( "this.tools", JSON.stringify(this.tools) );  
      console.log(
        "Connected to server with tools:",
        this.tools.map(({ name }) => name),
      );
    } catch (e) {
      console.log("Failed to connect to MCP server: ", e);
      throw e;
    }
  }

  async processQuery(query: string) {
    /**
     * Process a query using Claude and available tools
     *
     * @param query - The user's input query
     * @returns Processed response as a string
     */
    // Chat Details
    const chatRequest: requests.ChatRequest = {
        chatDetails: {
            compartmentId: env.TF_VAR_compartment_ocid,
            servingMode: servingMode,
            chatRequest: {
                message: query,
                apiFormat: "COHERE",
                maxTokens: 2000,
                temperature: 0,
                tools: this.tools,
            }
        },
        retryConfiguration: NoRetryConfigurationDetails
    };
    console.log( "chatRequest", JSON.stringify(chatRequest) );    
    const response = await this.llm.chat(chatRequest);
    const messages: Object[] = [
      {
        role: "user",
        content: query,
      },
    ];    

    // Process response and handle tool calls
    const finalText = [];

    finalText.push(response.text);
    if ( response.hasAttribute("toolCalls") ) {
      for (const toolCall of response.toolCalls) { 
        console.log(toolCall); 
        // Execute tool call
        const toolName = toolCall.name;
        const toolArgs = toolCall.parameters as { [x: string]: unknown } | undefined;
        finalText.push(
          `[Calling tool ${toolName} with args ${JSON.stringify(toolArgs)}]`,
        );
        const result = await this.mcp.callTool({
          name: toolName,
          arguments: toolArgs,
        });
        finalText.push(`[Calling tool done]`);

        // Continue conversation with tool results
        messages.push({
          role: "user",
          content: result.content as string,
        });
      }
    }

    return finalText.join("\n");
  }

  async chatLoop() {
    /**
     * Run an interactive chat loop
     */
    const rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout,
    });

    try {
      console.log("\nMCP Client Started!");
      console.log("Type your queries or 'quit' to exit.");

      while (true) {
        const message = await rl.question("\nQuery: ");
        if (message.toLowerCase() === "quit") {
          break;
        }
        console.log("\n" + message);        
        const response = await this.processQuery(message);
        console.log("\n" + response);
      }
    } catch (e) {
      console.log("Error: ", e);
      console.log(e.stack);
    } finally {
      rl.close();
    }
  }

  async cleanup() {
    /**
     * Clean up resources
     */
    await this.mcp.close();
  }
}

async function main() {
  if (process.argv.length < 3) {
    console.log("Usage: node build/index.js <path_to_server_script>");
    return;
  }
  const mcpClient = new MCPClient();
  try {
    await mcpClient.initLLM();
    await mcpClient.connectToServer(process.argv[2]);
    await mcpClient.chatLoop();
  } finally {
    await mcpClient.cleanup();
    process.exit(0);
  }
}

main();

File: mcp-client-python/oci_mcp_client.py
import asyncio
from typing import Optional
from contextlib import AsyncExitStack

from mcp import ClientSession, StdioServerParameters
from mcp.client.stdio import stdio_client
from dotenv import load_dotenv

import oci
import os
import traceback

load_dotenv()  # load environment variables from .env

# OCI
signer = oci.auth.signers.InstancePrincipalsSecurityTokenSigner()

class MCPClient:
    def __init__(self):
        # Initialize session and client objects
        self.session: Optional[ClientSession] = None
        self.exit_stack = AsyncExitStack()

    def llm_chat( self, query, chat_history, tools ):
        # OCI Signer
        region = os.getenv("TF_VAR_region")

        # #Tool definitions
        # tool1 = CohereTool()
        # tool1.name = "alarm_history"
        # tool1.description = "The tool will help you find information related to specific alarm history in Oracle Cloud within the specified time"
        # tool1.parameter_definitions = {
        #     "name": alarm_history_param
        # }        

        generative_ai_inference_client = oci.generative_ai_inference.GenerativeAiInferenceClient(
            config={}, 
            service_endpoint="https://inference.generativeai."+region+".oci.oraclecloud.com", 
            retry_strategy=oci.retry.NoneRetryStrategy(), 
            timeout=(10,240),
            signer=signer
        )
        chat_detail = oci.generative_ai_inference.models.ChatDetails()
        chat_request = oci.generative_ai_inference.models.CohereChatRequest()
        chat_request.message = query    
        chat_request.max_tokens = 4000
        chat_request.temperature = 1
        chat_request.chat_history = chat_history
        chat_request.frequency_penalty = 0
        chat_request.top_p = 0.75
        chat_request.top_k = 0
        chat_detail.serving_mode = oci.generative_ai_inference.models.OnDemandServingMode(
            model_id = os.getenv("TF_VAR_genai_cohere_model")
        )
        chat_detail.chat_request = chat_request
        chat_detail.compartment_id = os.getenv("TF_VAR_compartment_ocid")
     
        if tools:
            print( "tools:" + str(tools) )
            chat_tools = []
            for tool in tools:
                print( "tool:" + str(tool) )  
                params = {}
                if tool.get("input_schema"):
                    for key, value in tool["input_schema"]["properties"].items():
                        # Access key and value
                        print(f"Param: {key}: {value}")
                        params[key]= {
                            "description": key,
                            "type": value["type"],
                            "isRequired": False                            
                        }                        
                chat_tools.append( 
                    { 
                        "name": tool["name"], 
                        "description": tool["description"], 
                        "parameterDefinitions": params 
                    }
                )  
            print( "chat_tools:" + str(chat_tools) )
            chat_request.tools = chat_tools   

        print("-- chat_detail")
        print(vars(chat_detail))
        response = generative_ai_inference_client.chat(chat_detail)
        # Print result
        print("-- response")
        print(vars(response))
        return response.data.chat_response

    async def connect_to_server(self, server_script_path: str):
        """Connect to an MCP server
        
        Args:
            server_script_path: Path to the server script (.py or .js)
        """
        is_python = server_script_path.endswith('.py')
        is_js = server_script_path.endswith('.js')
        if not (is_python or is_js):
            raise ValueError("Server script must be a .py or .js file")
            
        command = "python" if is_python else "node"
        server_params = StdioServerParameters(
            command=command,
            args=[server_script_path],
            env=None
        )
        
        stdio_transport = await self.exit_stack.enter_async_context(stdio_client(server_params))
        self.stdio, self.write = stdio_transport
        self.session = await self.exit_stack.enter_async_context(ClientSession(self.stdio, self.write))
        
        await self.session.initialize()
        
        # List available tools
        response = await self.session.list_tools()
        tools = response.tools
        print("\nConnected to server with tools:", [tool.name for tool in tools])

    async def process_query(self, query: str) -> str:
        """Process a query using Claude and available tools"""
        response = await self.session.list_tools()
        available_tools = [{ 
            "name": tool.name,
            "description": tool.description,
            "input_schema": tool.inputSchema
        } for tool in response.tools]

        chat_history = [] 
        response = self.llm_chat(
            query = query,
            chat_history = chat_history,
            tools= available_tools
        )
        chat_history.append (
            {
                "role": "user",
                "content": query
            }
        )
        # Process response and handle tool calls
        final_text = []
        final_text.append(response.text)

        if response.toolsCalls:
            for tool_call in response.toolsCalls:
                tool_name = tool_call.name
                tool_args = tool_call.parameters
                
                # Execute tool call
                final_text.append(f"[Calling tool {tool_name} with args {tool_args}]")
                result = await self.session.call_tool(tool_name, tool_args)

                # Continue conversation with tool results
                if hasattr(content, 'text') and content.text:
                    chat_history.append({
                      "role": "CHATBOT",
                      "content": content.text
                    })
                chat_history.append({
                    "role": "USER", 
                    "content": result.content
                })

                # Get next response from Claude
                response = self.llm_chat(
                    query=result.content, 
                    chat_history=chat_history,
                    tools=None
                )

                final_text.append(response.content[0].text)

        return "\n".join(final_text)

    async def chat_loop(self):
        """Run an interactive chat loop"""
        print("\nMCP Client Started!")
        print("Type your queries or 'quit' to exit.")
        
        while True:
            try:
                query = input("\nQuery: ").strip()
                
                if query.lower() == 'quit':
                    break
                    
                response = await self.process_query(query)
                print("\n" + response)
                    
            except Exception as e:
                print(f"\nError: {str(e)}")
                print(traceback.format_exc())   
    
    async def cleanup(self):
        """Clean up resources"""
        await self.exit_stack.aclose()

async def main():
    if len(sys.argv) < 2:
        print("Usage: python client.py <path_to_server_script>")
        sys.exit(1)
        
    client = MCPClient()
    try:
        await client.connect_to_server(sys.argv[1])
        await client.chat_loop()
    finally:
        await client.cleanup()

if __name__ == "__main__":
    import sys
    asyncio.run(main())




